platform-dependent-components/problem-solver/cxx/sections-module/agent/get_decomposition_agent.cpp

```diff
@@ -17,75 +16,80 @@ using namespace utils;
-  ScAddr const langAddr =
-      IteratorUtils::getAnyByOutRelation(&m_memoryCtx, questionNode, scAgentsCommon::CoreKeynodes::rrel_3);
-  ScAddr decompositionAddr =
-      IteratorUtils::getAnyByOutRelation(&m_memoryCtx, questionNode, scAgentsCommon::CoreKeynodes::rrel_4);
-
-  if (!m_memoryCtx.IsElement(subjDomainAddr))
+
+  std::tuple<ScAddr, ScAddr, ScAddr, ScAddr> const Tresult = action.GetArguments<4>();
+  ScAddr const subjDomainAddr = std::get<0>(Tresult);
+  ScAddr const levelAddr = std::get<1>(Tresult);
+  ScAddr const langAddr = std::get<2>(Tresult);
+  ScAddr decompositionAddr = std::get<3>(Tresult);
+
+  if (!m_context.IsElement(subjDomainAddr))
   {
     SC_LOG_ERROR("GetDecompositionAgent: subject domain node not found.");
-    utils::AgentUtils::finishAgentWork(&m_memoryCtx, questionNode, false);
-    return SC_RESULT_ERROR_INVALID_PARAMS;
+    return action.FinishUnsuccessfully();
   }
-  if (m_memoryCtx.IsElement(levelAddr))
-    m_memoryCtx.GetLinkContent(levelAddr, level);
-  if (!m_memoryCtx.IsElement(langAddr))
+  if (m_context.IsElement(levelAddr))
+    m_context.GetLinkContent(levelAddr, level);
+  if (!m_context.IsElement(langAddr))
   {
     SC_LOG_ERROR("GetDecompositionAgent: language node not found.");
-    utils::AgentUtils::finishAgentWork(&m_memoryCtx, questionNode, false);
-    return SC_RESULT_ERROR_INVALID_PARAMS;
+    return action.FinishUnsuccessfully();
   }
-  if (!m_memoryCtx.IsElement(decompositionAddr))
+  if (!m_context.IsElement(decompositionAddr))
   {
     decompositionAddr = SectionsKeynodes::nrel_section_decomposition;
     SC_LOG_DEBUG(
         "GetDecompositionAgent: decomposition relation node not found. By default, "
-        << m_memoryCtx.HelperGetSystemIdtf(decompositionAddr) << " is used.");
+        << m_context.HelperGetSystemIdtf(decompositionAddr) << " is used.");
   }
 
-  ScAddr answerLink = m_memoryCtx.CreateLink();
+  ScAddr answerLink = m_context.CreateLink();
 
   ScAddrVector decomposition = GetDecomposition(subjDomainAddr, decompositionAddr);
   json answerJSON{
       {CommonUtils::getAddrHashString(subjDomainAddr),
-       {{"idtf", CommonUtils::getMainIdtf(&m_memoryCtx, subjDomainAddr, {langAddr})},
+       {{"idtf", CommonUtils::getMainIdtf(&m_context, subjDomainAddr, {langAddr})},
         {"decomposition", GetJSONDecomposition(decomposition, level, langAddr, decompositionAddr)},
         {"position", 0}}}};
 
   std::string answerJSONContent = answerJSON.dump();
   SC_LOG_INFO("Result decomposition: " << answerJSONContent);
-  m_memoryCtx.SetLinkContent(answerLink, answerJSONContent);
+  m_context.SetLinkContent(answerLink, answerJSONContent);
 
   ScAddrVector answerElements = {answerLink};
 
-  utils::AgentUtils::finishAgentWork(&m_memoryCtx, questionNode, answerElements, true);
+  ScStructure result = m_context.GenerateStructure();
+  for (auto const & answerElement : answerElements)
+    result << answerElement;
```

Comment:
and here there is no need of creating vector with 1 element to iterate trough it